test/e2e/clusterctl_upgrade.go

```diff
@@ -388,6 +391,11 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			kubernetesVersion = input.E2EConfig.GetVariable(KubernetesVersion)
 		}
 		controlPlaneMachineCount := ptr.To[int64](1)
+
+		// Don't set controlPlaneMachineCount to 0 to satisfy the validation in the cluster template.
+		if input.ControlPlaneMachineCount != nil && input.ControlPlaneMachineCount != ptr.To[int64](0) {
```

Comment:
I wonder if this comparison works. Are we comparing pointers? Are they actually ever equal?

Reply:
Whoops, fixed!